In [53]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import datetime as dt
%matplotlib inline

print ('numpy version:', np.__version__, 'pandas version:', pd.__version__)

numpy version: 1.21.5 pandas version: 1.3.4


### READ DATA

In [3]:
df = pd.read_excel('vesselData.xlsx')
df

eta                     ata                     atd  \
0     2017-09-19 00:00:00+00  2017-09-19 00:00:00+00  2017-09-22 00:00:00+00   
1     2017-10-02 00:00:00+00  2017-10-02 00:00:00+00  2017-10-03 00:00:00+00   
2     2017-09-30 00:00:00+00  2017-09-30 00:00:00+00  2017-10-01 00:00:00+00   
3     2017-10-02 00:00:00+00  2017-10-02 00:00:00+00  2017-10-03 00:00:00+00   
4     2017-10-02 00:00:00+00  2017-10-02 00:00:00+00  2017-10-02 00:00:00+00   
...                      ...                     ...                     ...   
8203  2017-11-03 00:00:00+00  2017-11-03 00:00:00+00  2017-11-04 00:00:00+00   
8204  2017-11-04 00:00:00+00  2017-11-04 00:00:00+00  2017-11-06 00:00:00+00   
8205  2017-11-08 00:00:00+00  2017-11-07 00:00:00+00  2017-11-11 00:00:00+00   
8206  2017-11-10 00:00:00+00  2017-11-10 00:00:00+00  2017-11-10 00:00:00+00   
8207  2017-11-10 00:00:00+00  2017-11-08 00:00:00+00  2017-11-10 00:00:00+00   

      vesseldwt  vesseltype  discharge1  load1  discharge2  load2  discharge3  \
0      109290.0         5.0         0.0    0.0         0.0    0.0     90173.0   
1       67170.0         3.0         0.0    0.0         0.0    0.0         0.0   
2       67737.0         3.0         0.0    0.0         0.0    0.0         0.0   
3       43600.0         3.0         0.0    0.0         0.0    0.0         0.0   
4        9231.0         3.0         0.0    0.0         0.0    0.0         0.0   
...         ...         ...         ...    ...         ...    ...         ...   
8203     9587.0         3.0         0.0    0.0         0.0    0.0         0.0   
8204     9654.0         3.0         0.0    0.0         0.0    0.0         0.0   
8205     4726.0         5.0         0.0    0.0         0.0    0.0         0.0   
8206    13320.0         3.0         0.0    0.0         0.0    0.0         0.0   
8207    11020.0         3.0         0.0    0.0         0.0    0.0         0.0   

      ...   load4                                     stevedorenames  \
0     ...     0.0                                      Stevedore_104   
1     ...     0.0                                      Stevedore_109   
2     ...     0.0                                       Stevedore_57   
3     ...     0.0                                       Stevedore_57   
4     ...     0.0                                       Stevedore_98   
...   ...     ...                                                ...   
8203  ...     0.0                                       Stevedore_64   
8204  ...     0.0  Stevedore_110,Stevedore_57,Stevedore_99,Steved...   
8205  ...  3051.0  Stevedore_89,Stevedore_79,Stevedore_75,Stevedo...   
8206  ...     0.0                                       Stevedore_46   
8207  ...     0.0  Stevedore_28,Stevedore_99,Stevedore_57,Stevedo...   

      hasnohamis             earliesteta               latesteta traveltype  \
0            NaN  2017-09-19 00:00:00+00  2017-09-19 00:00:00+00    ARRIVAL   
1            NaN  2017-10-02 00:00:00+00  2017-10-02 00:00:00+00    ARRIVAL   
2            NaN  2017-09-30 00:00:00+00  2017-09-30 00:00:00+00    ARRIVAL   
3            NaN  2017-10-02 00:00:00+00  2017-10-02 00:00:00+00    ARRIVAL   
4            NaN  2017-10-02 00:00:00+00  2017-10-02 00:00:00+00    ARRIVAL   
...          ...                     ...                     ...        ...   
8203         NaN  2017-11-02 00:00:00+00  2017-11-03 00:00:00+00    ARRIVAL   
8204         NaN  2017-11-04 00:00:00+00  2017-11-05 00:00:00+00      SHIFT   
8205         NaN  2017-11-07 00:00:00+00  2017-11-10 00:00:00+00      SHIFT   
8206         NaN  2017-11-10 00:00:00+00  2017-11-10 00:00:00+00    ARRIVAL   
8207         NaN  2017-11-08 00:00:00+00  2017-11-10 00:00:00+00      SHIFT   

     previousportid nextportid  isremarkable  vesselid  
0             981.0      731.0             f    2242.0  
1              19.0       15.0             f    5462.0  
2              19.0       19.0             f    5251.0  
3              15.0       18.0             f    5

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8208 entries, 0 to 8207
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   eta             8208 non-null   object 
 1   ata             8208 non-null   object 
 2   atd             8208 non-null   object 
 3   vesseldwt       8206 non-null   float64
 4   vesseltype      8208 non-null   float64
 5   discharge1      8208 non-null   float64
 6   load1           8208 non-null   float64
 7   discharge2      8208 non-null   float64
 8   load2           8208 non-null   float64
 9   discharge3      8208 non-null   float64
 10  load3           8208 non-null   float64
 11  discharge4      8208 non-null   float64
 12  load4           8208 non-null   float64
 13  stevedorenames  8206 non-null   object 
 14  hasnohamis      0 non-null      float64
 15  earliesteta     8208 non-null   object 
 16  latesteta       8208 non-null   object 
 17  traveltype      8208 non-null   o

### INVESTIGATE TARGET VARIABLES

In [28]:
targets_df = df[['load1', 'discharge1','load2', 'discharge2', 'load3', 'discharge3', 'load4', 'discharge4']]
targets_df

load1  discharge1  load2  discharge2  load3  discharge3   load4  \
0       0.0         0.0    0.0         0.0    0.0     90173.0     0.0   
1       0.0         0.0    0.0         0.0    0.0         0.0     0.0   
2       0.0         0.0    0.0         0.0    0.0         0.0     0.0   
3       0.0         0.0    0.0         0.0    0.0         0.0     0.0   
4       0.0         0.0    0.0         0.0    0.0         0.0     0.0   
...     ...         ...    ...         ...    ...         ...     ...   
8203    0.0         0.0    0.0         0.0    0.0         0.0     0.0   
8204    0.0         0.0    0.0         0.0    0.0         0.0     0.0   
8205    0.0         0.0    0.0         0.0    0.0         0.0  3051.0   
8206    0.0         0.0    0.0         0.0    0.0         0.0     0.0   
8207    0.0         0.0    0.0         0.0    0.0         0.0     0.0   

      discharge4  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  
...          ...  
8203         0.0  
8204         0.0  
8205      3537.0  
8206         0.0  
8207         0.0  

[8208 rows x 8 columns]

In [12]:
targets_df.describe()

load1     discharge1         load2     discharge2  \
count   8208.000000    8208.000000   8208.000000    8208.000000   
mean      60.754386    1732.712841     19.387549    1168.698221   
std     1325.463945   16297.994648    665.333248   11331.135301   
min        0.000000       0.000000      0.000000       0.000000   
25%        0.000000       0.000000      0.000000       0.000000   
50%        0.000000       0.000000      0.000000       0.000000   
75%        0.000000       0.000000      0.000000       0.000000   
max    41761.000000  204304.000000  43639.000000  189933.000000   

               load3     discharge3          load4     discharge4  
count    8208.000000    8208.000000    8208.000000    8208.000000  
mean       44.745492    4791.331871    1509.046784    1820.782407  
std      3071.717973   25366.035772   11936.519585   10034.936858  
min         0.000000       0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000       0.000000  
50%         0.000000       0.000000       0.000000       0.000000  
75%         0.000000       0.000000       0.000000       0.000000  
max    271251.000000  299647.000000  293449.000000  183837.000000

In [30]:
for i in range(1,5):
    print (f'load{i}',targets_df.loc[targets_df[f'load{i}'] > 0, f'load{i}'].count())
print('-----')
for i in range(1,5):
    print (f'discharge{i}',targets_df.loc[targets_df[f'discharge{i}'] > 0, f'discharge{i}'].count())

load1 40
load2 21
load3 8
load4 446
-----
discharge1 138
discharge2 112
discharge3 378
discharge4 558


### INITIAL INSIGHTS
Cargo type 4 (petrol) is the least type with missing load and discharge data (558 valid entries).
Based on a quick search https://bulkcarrierguide.com/cargo.html, I made an assumption: <br>
The handling of ore & coal might be similar (soild bulk cargo), also oil & petrol might have similar handling routines (safety, regulations, transportation, etc). <br>

Therefore:<br>
Attempting to concatenate load1 + load2 --> load_solid <br>
Attempting to concatenate load3 + load4 --> load_liquid <br>
Attempting to concatenate discharge1 + discharge2 --> discharge_solid <br>
Attempting to concatenate discharge3 + discharge4 --> discharge_liquid <br>

This is to maximize the supervision signal to train our predictive model.

### SOLUTION OVERVIEW: <br>
Train one load model to predict load for both solid and liquid cargo.<br>
Train one discharge model to predict discharge for both solid and liquid cargo<br>

In [64]:
valid_load1 = targets_df.loc[targets_df['load1'] > 0, 'load1']
valid_load2 = targets_df.loc[targets_df['load2'] > 0, 'load2']
load_solid = pd.concat([valid_load1, valid_load2] )

valid_load3 = targets_df.loc[targets_df['load3'] > 0, 'load3']
valid_load4 = targets_df.loc[targets_df['load4'] > 0, 'load4']
load_liquid = pd.concat([valid_load3, valid_load4] )
load_liquid


valid_disc1 = targets_df.loc[targets_df['discharge1'] > 0, 'discharge1']
valid_disc2 = targets_df.loc[targets_df['discharge2'] > 0, 'discharge2']
disc_solid= pd.concat([valid_disc1, valid_disc2] )
disc_solid


valid_disc3 = targets_df.loc[targets_df['discharge3'] > 0, 'discharge3']
valid_disc4 = targets_df.loc[targets_df['discharge4'] > 0, 'discharge4']
disc_liquid= pd.concat([valid_disc3, valid_disc4] )
disc_liquid 


0        90173.0
8        14044.0
27      131231.0
36       99862.0
50      102817.0
          ...   
8131      5104.0
8188     15980.0
8191      4196.0
8194     28127.0
8205      3537.0
Length: 936, dtype: float64

-----
### Feature selection & Engineering

1. construct a binary variable ```bulk_type``` to define the cargo type {'solid': 0, 'liquid': 1}
2. Isolate a subset of columns from the dataset as a candidate feature set
3. construct a ```process_time``` feature as the difference in days from the 2 dates ```atd ``` - ```ata```

In [60]:
# construct process_time feature
df['ata'] = pd.to_datetime(df['ata']).dt.date
df['atd'] = pd.to_datetime(df['atd']).dt.date
df['process_time'] = (df['atd'] - df['ata']).dt.days
df['process_time']

0       3
1       1
2       1
3       1
4       0
       ..
8203    1
8204    2
8205    4
8206    0
8207    2
Name: process_time, Length: 8208, dtype: int64

In [70]:
# construct the bulk_type feature
load = pd.DataFrame(columns= ['load', 'bulk_type'])
discharge = pd.DataFrame(columns= ['discharge', 'bulk_type'])
discharge

Empty DataFrame
Columns: [discharge, bulk_type]
Index: []

In [69]:
feature_candidates = ['vesseldwt', 'vesseltype', 'traveltype', 'isremarkable', 'process_time', 'bulk_type']

### FIT AN SVM/REGRESSION MODEL HERE :)
A load model and a discharge model